File: bitcoin_stockmem/pipeline/step4_reason.py
"""Step 4: Reflection Generation (LLM_reason).

Generates causal analysis linking event sequences to actual price movements.
Used during training and online learning (after test predictions are resolved).
"""

from __future__ import annotations

import logging

from config import WINDOW_SIZE
from llm.gemini_client import GeminiClient
from llm.prompts import REASON_PROMPT
from llm.response_parser import parse_reason_result
from memory.event_memory import build_event_series, format_series_for_prompt
from memory.reflection_memory import store_reflection

logger = logging.getLogger(__name__)


def generate_reflection(
    client: GeminiClient,
    date: str,
    asset: str,
    price_direction: str,
    price_change_pct: float | None = None,
    source: str = "train",
) -> int | None:
    """Generate a reflection for a given date and store it.

    Args:
        client: Gemini client.
        date: The analysis date (end of event window; prediction targets date+1).
        asset: "BTC" or "ETH".
        price_direction: Actual direction ("up" or "down").
        price_change_pct: Actual return percentage.
        source: "train" or "online".

    Returns:
        Reflection ID or None on failure.
    """
    # Build event series
    dates, events_per_day = build_event_series(date, WINDOW_SIZE, asset)
    information = format_series_for_prompt(dates, events_per_day, include_delta_info=True)

    if not information.strip():
        logger.warning(f"No events found for reflection on {date}/{asset}")
        return None

    prompt = REASON_PROMPT.format(
        asset=asset,
        information=information,
        price_change=price_direction,
    )

    try:
        result = client.generate_json(prompt)
        parsed = parse_reason_result(result)
    except Exception as e:
        logger.warning(f"Reflection generation failed for {date}/{asset}: {e}")
        return None

    # Store reflection
    ref_id = store_reflection(
        date=date,
        asset=asset,
        window_start=dates[0],
        window_end=dates[-1],
        price_direction=price_direction,
        reason=parsed.reason,
        key_events=parsed.key_events,
        price_change_pct=price_change_pct,
        source=source,
    )

    logger.info(f"Reflection {ref_id} stored for {date}/{asset} ({source})")
    return ref_id


def run_reflection_generation(
    client: GeminiClient,
    labels_df,
    asset: str,
    source: str = "train",
) -> list[int]:
    """Generate reflections for all labelled dates.

    Args:
        client: Gemini client.
        labels_df: DataFrame with columns 'date', 'label', 'next_return'.
        asset: "BTC" or "ETH".
        source: "train" or "online".

    Returns:
        List of reflection IDs.
    """
    ids = []
    for _, row in labels_df.iterrows():
        if row["label"] == "flat":
            continue
        ref_id = generate_reflection(
            client,
            date=row["date"],
            asset=asset,
            price_direction=row["label"],
            price_change_pct=row.get("next_return"),
            source=source,
        )
        if ref_id:
            ids.append(ref_id)
    return ids

File: bitcoin-stockmem-ts/src/pipeline/step4-reason.ts
/**
 * Step 4: Reflection Generation (LLM_reason).
 */

import { GeminiClient } from "../llm/gemini-client";
import { REASON_PROMPT, fillTemplate } from "../llm/prompts";
import { parseReasonResult } from "../llm/response-parser";
import { WINDOW_SIZE } from "../config";
import { buildEventSeries, formatSeriesForPrompt } from "../memory/event-memory";
import { storeReflection } from "../memory/reflection-memory";

export async function generateReflection(
  client: GeminiClient,
  date: string,
  asset: string,
  priceDirection: string,
  priceChangePct?: number,
  source = "train"
): Promise<number | null> {
  const { dates, eventsPerDay } = buildEventSeries(date, WINDOW_SIZE, asset);
  const information = formatSeriesForPrompt(dates, eventsPerDay, true);

  if (!information.trim()) {
    console.warn(`No events for reflection on ${date}/${asset}`);
    return null;
  }

  const prompt = fillTemplate(REASON_PROMPT, {
    asset,
    information,
    price_change: priceDirection,
  });

  try {
    const result = await client.generateJson(prompt);
    const parsed = parseReasonResult(result);

    const refId = storeReflection({
      date,
      asset,
      windowStart: dates[0],
      windowEnd: dates[dates.length - 1],
      priceDirection,
      reason: parsed["Reason for price movement"],
      keyEvents: parsed["Events causing the impact"],
      priceChangePct,
      source,
    });

    console.log(`Reflection ${refId} stored for ${date}/${asset} (${source})`);
    return refId;
  } catch (e: any) {
    console.warn(`Reflection failed for ${date}/${asset}: ${e.message}`);
    return null;
  }
}
